### PR TITLE
Refactor license tags

### PR DESCRIPTION
**Pull request by kopancek**
## Description

Today, license creation is purely based on human input. 

This PR tries to help in multiple ways:
- color-coding tags and showing a warning if a license was created with unknown tags - this is valuable for already created licenses
- returning a list of known tags from the backend, so that we are always in sync on which tags are supported
- improvements to the license creation form
  - refactored to use the new apollo client react hooks instead of rxjs
  - separated customer name and subscription plan into it's own input
  - tags are shown on the fly to the user and if they are invalid, we flag it as well

## Screenshots

Screenshot 2023-02-01 at 12 59 07

## Video

https://www.loom.com/share/7f390b711dd54d1fb11ab50740e969aa

## Test plan

Tested locally by running the following scenarios:
- create a license with invalid tags
- create a license with valid tags

## TODO
Update the related handbook page, so that we have up to date docs for CE and however creates the licenses.

## App preview:

- [Web](https://sg-web-milan-improve-license-tags-ux.onrender.com/search)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

### CODE REVIEW

**Comment by sg-e2e-regression-test-bob** [comment]
## Bundle size report 📦

| Initial size | Total size | Async size | Modules |
| --- | --- | --- | --- |
| 0.00% (0.00 kb) | 0.03% (+3.83 kb) | 0.03% (+3.83 kb) | 0.00% (0) |

Look at the [Statoscope report](https://console.cloud.google.com/storage/browser/_details/sourcegraph_reports/statoscope-reports/milan/improve_license_tags_ux/compare-report.html;tab=live_object) for a full comparison between the commits aa7f34c and b15987b or [learn more](https://docs.sourcegraph.com/dev/how-to/testing#bundlesize).


<details>
<summary>Open explanation</summary>

- `Initial size` is the size of the initial bundle (the one that is loaded when you open the page)
- `Total size` is the size of the initial bundle + all the async loaded chunks
- `Async size` is the size of all the async loaded chunks
- `Modules` is the number of modules in the initial bundle
</details>

File: client/web/src/enterprise/productSubscription/ProductLicenseTags.tsx [hunk 1]
```diff
@@ -1,6 +1,72 @@
 import React from 'react'
 
-import { Badge } from '@sourcegraph/wildcard'
+import { Alert, Badge, BadgeVariantType, Tooltip } from '@sourcegraph/wildcard'
+
+// TODO: instead of hardcoding, generate this list based on /enterprise/internal/licensing/data.go
```
Comment (mrnugget) [inline client/web/src/enterprise/productSubscription/ProductLicenseTags.tsx:5]: I think we could do that over the JS context? https://github.com/sourcegraph/sourcegraph/blob/main/cmd/frontend/internal/app/jscontext/jscontext.go

I don't think we should merge this with the list hardcoded in.

**Review by pjlast — APPROVED** [review]
I **really** like this. Thanks for doing this!

**Comment by kopancek** [comment]
Will merge this once the release branch is cut

**Review by vdavid — APPROVED** [review]
Nice, I didn't use the feature before, but the UX seems neat with your changes!

File: enterprise/internal/licensing/data.go [hunk 2]
```diff
@@ -84,6 +84,20 @@ const (
...
+	FeatureExplicitPermissionsAPI,
+	FeatureExtensionRegistry,
+	FeatureRemoteExtensionsAllowDisallow,
+	FeatureBranding,
+	FeatureCampaigns,
+	FeatureMonitoring,
+	FeatureBackupAndRestore,
+	FeatureCodeInsights,
+	&FeatureBatchChanges{},
+}
```
Comment (keegancsmith) [inline enterprise/internal/licensing/data.go:99]: @pjlast @kopancek should this list include `&FeaturePrivateRepositories`? Just noticed in main that it is not listed.

Comment (pjlast) [reply enterprise/internal/licensing/data.go:99]: It's not too relevant right now, but it is probably a good idea to add it there. the `AllFeatures` variable is only used to verify license tags, and the private repos feature isn't something that should be added to licenses at the moment. It's only used to gate the free licenses.

**Pull request merged**